psi4/src/psi4/libpsio/toclen.cc

```diff
@@ -64,53 +66,67 @@ size_t PSIO::toclen(size_t unit) {
     return (len);
 }
 
-size_t PSIO::rd_toclen(size_t unit) {
-    int errcod, stream;
-    psio_ud *this_unit;
-    size_t len;
-
-    this_unit = &(psio_unit[unit]);
-
-    /* Seek vol[0] to its beginning */
-    stream = this_unit->vol[0].stream;
-
-    errcod = SYSTEM_LSEEK(stream, 0L, SEEK_SET);
-
-    if (errcod == -1) psio_error(unit, PSIO_ERROR_LSEEK);
-
-    /* Read the value */
-
-    errcod = SYSTEM_READ(stream, (char *)&len, sizeof(size_t));
-
-    if (errcod != sizeof(size_t)) return (0); /* assume that all is well (see comments above) */
-
-    return (len);
-}
-
-void PSIO::wt_toclen(size_t unit, size_t len) {
-    int errcod, stream;
-    psio_ud *this_unit;
-
-    this_unit = &(psio_unit[unit]);
-
-    /* Seek vol[0] to its beginning */
-    stream = this_unit->vol[0].stream;
-
-    errcod = SYSTEM_LSEEK(stream, 0L, SEEK_SET);
-
+/// @brief Seek the stream of the vol[0] of a unit to its beginning
+/// @param unit : file unit number to rewind
+void PSIO::rewind_toclen(const size_t unit) {
+    if(!open_check(unit)) psio_error(unit, PSIO_ERROR_UNOPENED);
+    const auto stream = psio_unit[unit].vol[0].stream;
+    const auto errcod = SYSTEM_LSEEK(stream, 0L, SEEK_SET);
+    const auto sys_errno = errno;
     if (errcod == -1) {
-        ::fprintf(stderr, "Error in PSIO_WT_TOCLEN()!\n");
-        exit(_error_exit_code_);
+        std::string errmsg = "LSEEK failed. Error description from the OS: " + decode_errno(sys_errno);
+        errmsg += "\nCannot seek vol[0] to its beginning, unit ";
+        errmsg += std::to_string(unit) + ".\n";
+        psio_error(unit, PSIO_ERROR_LSEEK, errmsg);
     }
+}
 
-    /* Write the value */
-
-    errcod = SYSTEM_WRITE(stream, (char *)&len, sizeof(size_t));
+/// @brief Read the length of the TOC for a given unit directly from the file. Note that we do not exit if the read
+/// request of the toclen from the file fails. This is because the request may be to an new file for which the toclen
+/// has not yet been written. (We allow the user open files with status PSIO_OPEN_OLD even if they don't exist, because
+/// sometimes you can't know this in advance.)
+/// @param unit : file unit number to read TOC length from
+/// @return length of the TOC for a given unit
+size_t PSIO::rd_toclen(const size_t unit) {
+    if(!open_check(unit)) psio_error(unit, PSIO_ERROR_UNOPENED);
+    // Seek to the beginning
+    rewind_toclen(unit);
+    // Read the value
+    size_t len;
+    const auto stream = psio_unit[unit].vol[0].stream;
+    const auto errcod = SYSTEM_READ(stream, (char *)&len, sizeof(size_t));
+    const auto sys_errno = errno;
+    if (errcod != sizeof(size_t)) {
+#ifdef DEBUG
```

Comment:
Is this intended to occur only within the `#ifdef`?  The corresponding code in `wt_toclen()` isn't similarly blocked.

Reply:
Good point, it is something I was on the fence about.

The `read(...)` system call returns either the number of bytes read, or `-1` if an error occurs. The old code checks `if(bytes_read_expected==bytes_read_actual)` and returns 0 if that is false. This is justified by the original code as the following:

> Note that we do not exit if the read request of the toclen from the file fails. This is because the request may be to an new file for which the toclen has not yet been written. (We allow the user open files with status PSIO_OPEN_OLD even if they don't exist, because sometimes you can't know this in advance.)

So after reading that I chose to preserve the original behavior unless the `#ifdef` is active, but now I am not so sure, as the original behavior fails to guard against actual read errors.

Given how the code in the `#ifdef` block should only trigger on actual read errors it might be best to not only remove the `#ifdef`, but make it an error exit like the other failures.

Reply:
Yes, I'm pretty sure I wrote the original version of that comment when I wrote `libpsio` so long ago.  If the current version with the removal of the `#ifdef` runs correctly, then I'm happy to LGTM it.